### ISSUE TITLE
Add a `ViewRef.onMarkDirty(callback)` method as an enabler for custom Change Detection strategies + debug tools

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular#49882]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

**TL;DR: Adding a `ViewRef.onMarkDirty(callback)` method would allow library authors and researchers to provide or experiment with custom change detection strategies in advanced use cases. This could also allow implementing useful devtools.**

Thanks to Signals, Angular has a relatively fine-grained knowledge of which views should be marked for check.
This information is not currently shared with the developer. If it was propagated through a public API, this would be an enabler for different community initiatives:

#### 1. Custom Change Detection Strategies
Developers could implement custom change detection strategies per view by detaching the change detector and implementing their own logic.

```ts
viewRef.detach();
viewRef.onMarkDirty(() => customScheduler.scheduleChangeDetection(viewRef));
``` 

#### 2. DevTools

We can imagine new devtools tracking which components are marked dirty when a specific signal changes for example.

### Drawbacks

The main drawbacks of exposing such an API are:
- the abuse and misuse of `onMarkDirty()` and naive scheduling strategies that would schedule more change detections or manual renders than necessary... or even trigger some infinite loops.
- triggering change detection synchronously in the callback would read signals while "write" is propagated
(that said, this would throw an error, right?)

### When to call the callback?

Should the callback be triggered whenever `markViewDirty()` is called? Or do we need more fine-grained notifications to tell if an input changed, or a signal (might have) changed, or an event listener was triggered?

### Should it be exposed on `ChangeDetectorRef`?

The method could be exposed on `ChangeDetectorRef`.

### Proposed solution

Add a `ViewRef.onMarkDirty(callback: () => void)` method to notify developers when the view is marked dirty using `markViewDirty(view)`

whether it's caused by signals:
https://github.com/angular/angular/blob/e9dd7f00280b6d52fa2a21da793521db1a9d7a53/packages/core/src/render3/reactive_lview_consumer.ts#L32

or other means.

This could be combined with https://github.com/angular/angular/issues/49884

### Alternatives considered

Overriding `ReactiveLViewConsumer.onConsumerDependencyMayHaveChanged()` at runtime.

Cf. https://marmicode.io/blog/angular-signals-and-custom-render-strategies

This is fragile as it's hacking into internal APIs.

**Comment by atscott** [comment]
Closing for same reason as described in https://github.com/angular/angular/issues/49884#issuecomment-1511612293 (better zoneless support is already tracked as a larger project and we won't be tracking individual proposed solutions to the problem-space).
